lib/mix/lib/mix/utils.ex

```diff
@@ -29,17 +29,33 @@ defmodule Mix.Utils do
 
   """
   def mix_config do
-    mix_home_xdg_lookup("XDG_CONFIG_HOME")
+    mix_home_lookup(["MIX_HOME", {:system, :user_config}])
   end
 
-  defp mix_home_xdg_lookup(xdg) do
-    case {System.get_env("MIX_HOME"), System.get_env(xdg)} do
-      {directory, _} when is_binary(directory) -> directory
-      {nil, directory} when is_binary(directory) -> Path.join(directory, "mix")
-      {nil, nil} -> Path.expand("~/.mix")
+  defp mix_home_lookup([]), do: Path.expand("~/.mix")
+
+  defp mix_home_lookup([{:system, type} | rest]) do
+    case xdg?() do
+      true -> xdg_dir(type)
+      _ -> mix_home_lookup(rest)
+    end
+  end
+
+  defp mix_home_lookup([env | rest]) when is_binary(env) do
+    case System.get_env(env) do
+      nil -> mix_home_lookup(rest)
+      dir -> dir
     end
   end
 
+  defp xdg_dir(type), do: :filename.basedir(type, "mix", %{os: :linux})
+
+  defp xdg? do
+    File.dir?(xdg_dir(:user_config)) or
+      File.dir?(xdg_dir(:user_data)) or
+      Enum.any?(~w[XDG_CONFIG_HOME XDG_DATA_HOME], &System.get_env/1)
```

Comment:
I don't think we should use XDG for config if only XDG_DATA_HOME is set and vice versa. I feel like it would be confusing if setting one variable changes another's default.

Reply:
I think it would be more confusing when I would have `~/.config/mix` used, but at the same time it would still create `~/.mix` for data. Now it has only 2 modes:

- XDG-like
- Legacy

Instead of 4:

- Fully XDG-like
- Configuration XDG-like and legacy
- Data XDG-like and legacy
- Legacy

Reply:
But I don't see why you wouldn't set both variables if you want to use XDG directories. You could also argue that your suggestion is also 4 scenarios, with more magic happening in background that would be more to document:

- Fully XDG-like
- Configuration XDG-like and fully XDG data
- Data XDG-like and fully XDG configuration
- Legacy

It's more straightforward to have a single default with overrides, instead of overrides that also change some of the defaults.

Reply:
The point is that XDG Base Directory Specification defines defaults for each of the variables, i.e. if `XDG_CONFIG_HOME` is not set it **should** be treated as it has value `$HOME/.config`. So this is 100% conformant to the spec, so I fully disagree with the reasoning that it has 4 different behaviours. Example of situation when user would set only one variable:

You want to manage your dotfiles using VCS, so you create `~/dotfiles` and init your VCS there. Then you just define `XDG_CONFIG_HOME=$HOME/dotfiles/config` and leave rest as is, as you are fine with the defaults (which mean `XDG_DATA_HOME` is `$HOME/.local/share`).

Reply:
We are not fully conformant to the spec, that's why you created this PR, and even after this PR we still won't be fully conformant to the spec. So the question here is what is least confusing to users and what is easiest to explain, not what is most conformant to the spec.

Having a single default that doesn't change by setting other environment variables is the least confusing. If you write the docs for these changes I think you will also notice that this behavior is easier to explain.

Reply:
Nah… I do not think it was easier to explain old behaviour. If anything I think the old one was more confusing, but I may be a little biased as I know how XDG BDS works. However I would say that anyone who will opt-in to the XDG-like behaviour will also know how it works.

Reply:
> However I would say that anyone who will opt-in to the XDG-like behavior will also know how it works.

In this case it would be how mix implements parts of XDG, not how XDG works. XDG does not say how things should work if only one variable is set and the defaults are not being followed. This means it's up to mix to decide how this should work.

Additionally, making this change assumes that some people will only set one of the XDG variables, under this assumption we would make a breaking change for all those users.

Reply:
> XDG does not say how things should work if only one variable is set and the defaults are not being followed.

I do not know how about you, but for me these lines are pretty clear:

> `$XDG_DATA_HOME` defines the base directory relative to which user specific data files should be stored. If `$XDG_DATA_HOME` is either not set or empty, a default equal to `$HOME/.local/share` should be used.
>
> `$XDG_CONFIG_HOME` defines the base directory relative to which user specific configuration files should be stored. If `$XDG_CONFIG_HOME` is either not set or empty, a default equal to `$HOME/.config` should be used.

My point is that either we enable compatibility layer full or no at all. I prefer to enable it fully if user opt-in into any of them.

> Additionally, making this change assumes that some people will only set one of the XDG variables, under this assumption we would make a breaking change for all those users.

I assume that people who want such behaviour set both variables because they are forced to do so, and in perfect world they will prefer to set none of them.

Reply:
> I do not know how about you, but for me these lines are pretty clear.

Yes, but we do not fully follow the specification you quoted, even after this PR, so I am not sure how it applies. Deciding on how, how much, and when to follow the specs is precisely what we are discussing here. We all know the specs, they have been quoted and linked to multiple times, what's up to decide is how we can enable users to use XDG directories without breaking things.

> I assume that people who want such behaviour set both variables because they are forced to do so, and in perfect world they will prefer to set none of them.

If that's the case let's keep the old behavior, since the change you are suggesting would make no difference to those users. This way we do not break things and users who know about XDG can continue setting both variables.

Reply:
> This way we do not break things and users who know about XDG can continue setting both variables.

- Let's do not migrate to IPv6 because there are users who still uses IPv4 and if someones uses IPv6 then they can use both addressing systems.
- Let's do not migrate from FAT32 because there are OSes who still supports only that FS, and if someone wants to store large file then they can always split it into smaller ones.
- Let's do not migrate to TLSv1.3 as there are still systems that do not support these and it will break these who old ciphers.

Sorry, but I cannot accept that as an argument. "Hey, just configure your environment like that, because I do not want to conform to specs" is no argument.

This suits there perfectly - https://xkcd.com/1172/

Reply:
> Sorry, but I cannot accept that as an argument. "Hey, just configure your environment like that, because I do not want to conform to specs" is no argument.

We are not going to conform to the specs anyway - we can't because it would be a breaking change. This much should be clear to everyone here.

Both solutions proposed here are arbitrary. You are both discussing which arbitrary solution is better.

That said, I wonder if setting MIX_XDG would be better. Since you have to set an environment variable, maybe it is best to set to one that you can simply set to `true` or `1`.

Reply:
The other option is to not support XDG at all, given that we don't conform to the spec. And honestly, not supporting it at all is easier than having to handle all these arbitrary discussions.

Reply:
@josevalim this can be a solution as well, but I would still check for the presence of the directory. In general I would go with:

- Load `$XDG_CONFIG_DIRS/mix/config.exs` files (`XDG_CONFIG_DIRS` is list of `:` separated directories in order of preference) if present
- Load `$XDG_CONFIG_HOME/mix/config.exs` if present and overwrite earlier options
- Load `$MIX_HOME/config.exs` if present and overwrite earlier options

And in similar vain treat the rest of the configuration files (escripts and archives). Then at some point in future add deprecation warning and move to XDG-conforming behaviour.

This gives additional feature of global configuration.

Reply:
Unfortunately there is no straight-forward migration path and the directory checking doesn't work generally either. This has been explained in other discussions too. To recap:

Think about the scenario where a company has a script for running Elixir on CI. The scripts checks for `~/.mix`. Because the script can run on a new machine, it won't have the XDG mix directory and it won't have the HOME mix either. But we still need to go to `~/.mix`, not doing so would be a breaking change.

So assuming we always have to write to `~/.mix`, given the above, warning for using `~/.mix` means every Elixir new user will get a warning, because that's where we need to write on clean installs.

The only scenario we could safely ignore `~/.mix` is if the `XDG_CONFIG_DIRS/mix` exists but not the `~/.mix` one. But honestly I wonder when would that ever happen? Also, it means that if you nuke your config directory, Mix would revert to writing to `~/.mix`.

I want to have this solved once for all.  We have three options:

1. Remove XDG support
2. Merge this PR as is (or using @ericmj's approach - to me both are equally arbitrary)
3. Opt-in with MIX_XDG=1 which IMO is easier than setting any of the XDG variables

If we can't reach a consensus, then my vote is 1.